[CodeCompletion] Show the enum type for enum element with assoc values

rintaro opened the pull request:
```
enum MyEnum {
    case null
    case str(String)
}
```

When completing elements for enum like this, the former shows `MyEnum`, but the latter shows `(String) -> MyEnum` as the display type. This is inconsistent with function/method call pattern which only shows the result type.

rdar://problem/48220244

rintaro committed:
[CodeCompletion] Show the enum type for enum element with assoc values

enum MyEnum {
    case null
    case str(String)
}

When completing elements for enum like this, the former shows `MyEnum`,
but the latter shows `(String) -> MyEnum`. This is inconsistent with
function call pattern which only shows the result type.

rdar://problem/48220244

rintaro commented:
@swift-ci Please smoke test

slavapestov commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -2667,17 +2667,19 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
+    if (auto FuncType = EnumType->getAs<AnyFunctionType>()) {
+      EnumType = FuncType->getResult();
+      if (auto FuncType = EnumType->getAs<FunctionType>())
         EnumType = FuncType->getResult();
```
The result type of an enum constructor is always the type of the enum itself. Why not just use that instead of looking at EED->getInterfaceType()?

slavapestov replied:
ie, just call `ED->getDeclaredInterfaceType()` where ED is your EnumDecl

rintaro replied:
Not strong reason. But we want to display bound generic types for parameters as well as the result enum type. To do that, I want to use `getTypeOfMember()` because it can substitute generic parameters all at once.

I added another commit for another commit for displaying bound generic types. Could you take a look?

rintaro committed:
[CodeCompletion] Show bound generic types for enum elements

We weren't displaying bound generic types for enum elements. It's
inconsistent with functions and initializiers.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test OS X platform
